#include "mainwindow.h"
#include "ui_mainwindow.h"
#include <JsonParser.h>
#include <adddevice_dialog.h>
#include <qtreewidget.h>
#include <QTreeWidgetItem>
#include <fstream>
#include <sstream>
#include <future>
#include <thread>
void MainWindow::ShowDevices()
{
    ui->listWidget->clear();
    QString path = "/home/wielenote/Documents/devices.json";
    auto file = JsonFile::ReadFile(path);
    auto doc = QJsonDocument::fromJson(file);
    auto doc_object = doc.object();

    auto list_of_devices = doc_object.keys();
    ui->listWidget->addItems(list_of_devices);
}
MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
{
    ui->setupUi(this);
    ui->tabWidget->setCurrentIndex(0);
    ui->label_Warning->setVisible(false);
    ShowDevices();
}

MainWindow::~MainWindow()
{
    delete ui;
}


void MainWindow::onResult(QNetworkReply *reply)
{
    auto er = reply->error();
    if(er == QNetworkReply::NoError)
    {
        QJsonDocument document = QJsonDocument::fromJson(reply->readAll());
        auto Object = document.object().value("cves");
        auto array = Object.toArray();
        auto currentObject = array.at(0).toObject();
        for(const auto& item : currentObject.keys())
        {

            qDebug() << item <<'-' << currentObject.value(item).toVariant().toString();
        }
    }
    reply->deleteLater();
}

void MainWindow::on_add_button_clicked()
{
    AddDevice_dialog dialog(this);
    this->hide();
    dialog.exec();
}

static std::vector<std::string> split(const std::string &s, char delim)
{
    std::vector<std::string> elems;
    std::stringstream ss(s);
    std::string item;
    while (std::getline(ss, item, delim)) {
        elems.push_back(item);
    }
    return elems;
}


void GetVulnerabilities(string path, string package_name)
{

}
//const char*
std::string exec(string cmd) {
    char buffer[128];
    std::string result = "";
    FILE* pipe = popen(cmd.data(), "r");
    if (!pipe) throw std::runtime_error("popen() failed!");
    {
        while (fgets(buffer, sizeof buffer, pipe) != NULL)
            result += buffer;
    }

    pclose(pipe);
    return result;
}

static bool IsVulnerable(const string& device_version,const string& vulner_version)
{
    stringstream ss;
    ss << "dpkg --compare-versions \"" << device_version << "\" \"ge\" \"" << vulner_version << "\"; echo $?";
    return exec(ss.str())[0] == '1'; // 1 - false & 0 - true
}

void MainWindow::on_tabWidget_currentChanged(int index)
{
    if (index == 2)
    {
            //ui->label_Warning->setVisible(true);
        auto a = IsVulnerable("0.6.55-0ubuntu12~20.04.5", "0.6.55-0ubuntu12~20.04.5");
        auto b = IsVulnerable("1.1", "1.2.1");
        qDebug() << b;
    }
    if(index == 1)
    {
        auto path = "/home/wielenote/build-QueryTest-Desktop_Qt_6_5_3_GCC_64bit-Debug/script_result_ZOS.txt";
        std::ifstream stream(path);
        string line;
        getline(stream,line);
        QString folder_path = "/home/wielenote/build-QueryTest-Desktop_Qt_6_5_3_GCC_64bit-Debug/vulnerabilities/";
        if (stream)
        {
            while(getline(stream,line))
            {
                auto sub_strs = split(line, ' ');
                if (sub_strs[0] == "zip")
                {
                    qDebug() << "123";
                }
                ui->treeWidget->setColumnCount(1);

                QTreeWidgetItem *res = new QTreeWidgetItem();
                int count = 0;
                int index_of_file = 0;
                do {
                    ++index_of_file;
                    auto file_path = folder_path+QString::fromStdString(sub_strs[0])+ "-" + QString::number(index_of_file) +".json";
                    auto a = JsonFile(file_path,sub_strs[0]);
                    auto val = a.GetVulnerabilities();
                    if  (count == 0)
                        count = a.GetCountFiles();
                    if (val.size() == 0)
                        break;


                    for (const auto& item : val)
                    {
                        QTreeWidgetItem *cve_root = new QTreeWidgetItem();
                        QTreeWidgetItem *cve_info_root1 = new QTreeWidgetItem();
                        QTreeWidgetItem *cve_info_root2 = new QTreeWidgetItem();
                        QTreeWidgetItem *cve_info_root_description = new QTreeWidgetItem();
                        QTreeWidgetItem *cve_info_root_description_value = new QTreeWidgetItem();
                        std::stringstream ss;
                        ss << std::fixed << std::setprecision(2) << item.GetCVSS();


                        cve_root->setText(0, QString::fromStdString(item.GetCVE_ID()));




                        if((item.GetStatus() == "released" && IsVulnerable(sub_strs[1], item.GetReleasedVersion())) ||
                            item.GetStatus() != "released" )
                            cve_root->setForeground(0,QBrush(Qt::red));





                        auto status = QString::fromStdString(item.GetStatus());
                        cve_info_root1->setText(0, "Status: " + status);
                        cve_info_root2->setText(0, "CVSS3: " + QString::fromStdString(ss.str()));
                        cve_info_root_description->setText(0, "Description");
                        cve_info_root_description_value->setText(0, QString::fromStdString(item.GetDescription()));
                        cve_root->addChild(cve_info_root1);
                        cve_root->addChild(cve_info_root2);

                        cve_info_root_description->addChild(cve_info_root_description_value);
                        cve_root->addChild(cve_info_root_description);

                        if (status != "Unknown")
                            res->addChild(cve_root);

                    }
                } while(index_of_file < count);

                if(res->childCount() != 0)
                {
                    QTreeWidgetItem *packge_root = new QTreeWidgetItem(ui->treeWidget);
                    packge_root->insertChildren(0,res->takeChildren());
                    ui->treeWidget->addTopLevelItem(packge_root);
                    packge_root->setText(0, QString::fromStdString(sub_strs[0]));
                }
            }


        }

    }
    else
    {
        ui->treeWidget->clear();
    }
}

bool RunScript(Ui::MainWindow* ui)
{
    QString path = "/home/wielenote/Documents/devices.json";
    auto file = JsonFile::ReadFile(path);
    auto doc = QJsonDocument::fromJson(file);
    auto doc_object = doc.object();
    auto name = ui->listWidget->currentItem()->text();
    auto device = doc_object.value(name).toObject();
    stringstream ss;
    ss << "/home/wielenote/CourseProject/VulnerabilityScanner/check_installed_packages.sh " << device.value("password").toVariant().toString().toStdString()
       << " " << device.value("login").toVariant().toString().toStdString() << " " <<
        device.value("ip_adress").toVariant().toString().toStdString();// ./check_installed_packages.sh
    qDebug() << QString::fromStdString(ss.str());
    system(ss.str().c_str());
    return true;
}
void MainWindow::on_pushButton_clicked()
{
    //int total = stoi(exec("wc -l /home/wielenote/build-QueryTest-Desktop_Qt_6_5_3_GCC_64bit-Debug/script_result_ZOS.txt")) - 1;
    //ui->progressBar->setMaximum(total);


    ui->label_Warning->setVisible(true);


    future<bool> ft = async(RunScript, ui);
    ft.wait();
    if (ft.get())
    {
        ui->label_Warning->setVisible(false);
    }
    //ui->label_Warning->setVisible(false);
    auto a = 0;
}


void MainWindow::on_scan_button_clicked()
{
    ui->label_Warning->setVisible(!ui->label_Warning->isVisible());
}

